bundles/org.openhab.binding.insteon/src/main/java/org/openhab/binding/insteon/internal/command/InsteonCommandExtension.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.insteon.internal.command;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.smarthome.io.console.Console;
+import org.eclipse.smarthome.io.console.extensions.AbstractConsoleCommandExtension;
+import org.eclipse.smarthome.io.console.extensions.ConsoleCommandExtension;
+import org.openhab.binding.insteon.internal.handler.InsteonNetworkHandler;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ *
+ * Console commands for the Insteon binding
+ *
+ * @author Rob Nielsen - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = ConsoleCommandExtension.class)
+public class InsteonCommandExtension extends AbstractConsoleCommandExtension {
+
+    private static final String DISPLAY_DEVICES = "display_devices";
+    private static final String DISPLAY_CHANNELS = "display_channels";
+    private static final String DISPLAY_LOCAL_DATABASE = "display_local_database";
+
+    public InsteonCommandExtension() {
+        super("insteon", "Interact with the Insteon integration.");
+    }
+
+    @Override
+    public void execute(String[] args, Console console) {
+        if (args.length > 0) {
+            InsteonNetworkHandler handler = InsteonNetworkHandler.getInsteonNetworkHandler();
```

Comment:
Don't retrieve a InsteonNetworkHandler instance statically. Instead you want to get it from OSGI.

You can add the following methods to this class.
```
@Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
public void setInsteonNetworkHandler(InsteonNetworkHandler handler){
//logic
}
public void unsetInsteonNetworkHandler(InsteonNetworkHandler handler){
//logic
}
```

And then in the handler factory add:
```
    private final Map<ThingUID, @Nullable ServiceRegistration<?>> serviceRegs = new HashMap<>();


    private synchronized void registerInsteonNetworkHandler(InsteonNetworkHandler handler) {
        this.serviceRegs.put(handler.getThing().getUID(), bundleContext.registerService(InsteonNetworkHandler.class.getName(),
                adapter, new Hashtable<String, Object>()));
    }

    @Override
    protected synchronized void removeHandler(ThingHandler thingHandler) {
        if (thingHandler instanceof InsteonNetworkHandler) {
            UID uid = thingHandler.getThing().getUID();
            ServiceRegistration<?> serviceReg = this.serviceRegs.remove(uid);
            if (serviceReg != null) {
                serviceReg.unregister();
            }
        }
    }
```

Just make sure to call `registerInsteonNetworkHandler` when you create the InsteonNetworkHandler instance.
The code may not work verbatim so you might have to adjust it to get it to work, but you should get the idea.

Reply:
Implemented as suggested

Reply:
implemented